File: Sources/YTools/Services/SearchCoordinator.swift
import Foundation
import YToolsModuleKit

struct RegisteredSearchModule: Sendable {
    let module: any YToolsModule
    let policy: ModuleResultPolicy
    let contentType: SearchContentType

    init(
        _ module: any YToolsModule,
        contentType: SearchContentType,
        allowedCapabilities: Set<ModuleCapability> = [],
        allowsPrivilegedActions: Bool = false
    ) {
        self.module = module
        self.contentType = contentType
        self.policy = ModuleResultPolicy(
            allowedCapabilities: allowedCapabilities,
            allowsPrivilegedActions: allowsPrivilegedActions
        )
    }
}

struct BackgroundSearchRequest: Sendable {
    let query: String
    let fileNavigationActive: Bool
    let showsHiddenFiles: Bool
    let fileNavigationSort: FileNavigationSort
    let fileNavigationSortAscending: Bool
    let fileNavigationFoldersFirst: Bool
    let enabledContentTypes: Set<SearchContentType>
    let applicationAliases: [String: String]
    let requestModules: [RegisteredSearchModule]
}

/// Owns every non-Spotlight query provider. All results—including trusted
/// built-ins—cross the same descriptor, capability, field and action policy.
actor SearchCoordinator {
    private let applications = ApplicationModule()
    private let fileNavigation = FileNavigationModule()
    private let standardModules: [RegisteredSearchModule]

    init(personalModules: [any YToolsModule] = [TextStatisticsModule()]) {
        standardModules = [
            RegisteredSearchModule(CalculatorModule(), contentType: .calculations),
            RegisteredSearchModule(UnitConversionModule(), contentType: .calculations),
            RegisteredSearchModule(SpellingModule(), contentType: .dictionary),
            RegisteredSearchModule(SettingsModule(), contentType: .systemTools)
        ] + personalModules.map { RegisteredSearchModule($0, contentType: .textTools) }
    }

    func prepare() async {
        await applications.prepare()
    }

    func search(_ request: BackgroundSearchRequest) async -> [LauncherResult] {
        guard !Task.isCancelled else { return [] }
        if request.fileNavigationActive {
            guard request.enabledContentTypes.contains(.files) else { return [] }
            let descriptor = ModuleDescriptor(
                id: "file-navigation",
                name: "文件导航",
                capabilities: [.localFileRead]
            )
            let results = fileNavigation.results(
                for: request.query,
                showsHiddenFiles: request.showsHiddenFiles,
                sort: request.fileNavigationSort,
                ascending: request.fileNavigationSortAscending,
                foldersFirst: request.fileNavigationFoldersFirst
            )
            return sanitize(
                results,
                descriptor: descriptor,
                policy: ModuleResultPolicy(allowedCapabilities: [.localFileRead])
            )
        }

        async let applicationResults = request.enabledContentTypes.contains(.applications)
            ? searchApplications(query: request.query, aliases: request.applicationAliases)
            : []
        async let moduleResults = searchModules(
            query: request.query,
            registrations: standardModules + request.requestModules,
            enabledContentTypes: request.enabledContentTypes
        )
        let combined = await applicationResults + moduleResults
        return Task.isCancelled ? [] : combined
    }

    private func searchApplications(query: String, aliases: [String: String]) async -> [LauncherResult] {
        let descriptor = ModuleDescriptor(
            id: "applications",
            name: "应用程序",
            capabilities: [.localFileRead]
        )
        let results = await applications.results(for: query, aliases: aliases)
        return sanitize(
            results,
            descriptor: descriptor,
            policy: ModuleResultPolicy(allowedCapabilities: [.localFileRead])
        )
    }

    private func searchModules(
        query: String,
        registrations: [RegisteredSearchModule],
        enabledContentTypes: Set<SearchContentType>
    ) async -> [LauncherResult] {
        await withTaskGroup(of: [LauncherResult].self, returning: [LauncherResult].self) { group in
            for registration in registrations {
                guard !Task.isCancelled else {
                    group.cancelAll()
                    break
                }
                guard enabledContentTypes.contains(registration.contentType) else { continue }
                let descriptor = registration.module.descriptor
                guard registration.policy.permits(descriptor) else { continue }
                let wasAdded = group.addTaskUnlessCancelled {
                    await Self.search(registration.module, query: query, policy: registration.policy)
                }
                if !wasAdded { break }
            }
            var combined: [LauncherResult] = []
            for await results in group {
                guard !Task.isCancelled else {
                    group.cancelAll()
                    return []
                }
                combined.append(contentsOf: results)
            }
            return combined
        }
    }

    private static func search(
        _ module: any YToolsModule,
        query: String,
        policy: ModuleResultPolicy
    ) async -> [LauncherResult] {
        do {
            let request = ModuleSearchRequest(query: query, maximumResults: 40)
            let results = try await module.search(request)
            guard !Task.isCancelled else { return [] }
            return results.prefix(request.maximumResults).compactMap {
                policy.sanitize($0, from: module.descriptor)
            }
        } catch {
            return []
        }
    }

    private func sanitize(
        _ results: [LauncherResult],
        descriptor: ModuleDescriptor,
        policy: ModuleResultPolicy
    ) -> [LauncherResult] {
        results.prefix(40).compactMap { policy.sanitize($0, from: descriptor) }
    }
}

File: src/YTools.Windows/Services/SearchCoordinator.cs
using YTools.Models;
using YTools.ModuleKit;
using YTools.Services.Modules;

namespace YTools.Services;

public sealed record RegisteredSearchModule(
    IYToolsModule Module,
    SearchContentType ContentType,
    ModuleResultPolicy Policy);

public sealed record BackgroundSearchRequest(
    string Query,
    bool FileNavigationActive,
    bool ShowsHiddenFiles,
    FileNavigationSort FileNavigationSort,
    bool FileNavigationSortAscending,
    bool FileNavigationFoldersFirst,
    IReadOnlySet<SearchContentType> EnabledContentTypes,
    IReadOnlyDictionary<string, string> ApplicationAliases,
    IReadOnlyList<string> CustomApplicationPaths,
    IReadOnlyList<RegisteredSearchModule> RequestModules,
    CancellationToken CancellationToken,
    long Generation = 0);

/// <summary>
/// Owns every non-file query provider. All results—including trusted
/// built-ins—cross the same descriptor, capability, field and action policy.
/// </summary>
public sealed class SearchCoordinator
{
    private readonly ApplicationIndexService _applications = new();
    private readonly FileNavigationModule _fileNavigation = new();
    private readonly IReadOnlyList<RegisteredSearchModule> _standardModules;

    public SearchCoordinator(
        SpellingService spelling,
        IReadOnlyList<IYToolsModule>? personalModules = null)
    {
        var standards = new List<RegisteredSearchModule>
        {
            new(new CalculatorModule(), SearchContentType.Calculations, new ModuleResultPolicy()),
            new(new UnitConversionModule(), SearchContentType.Calculations, new ModuleResultPolicy()),
            new(new SpellingModule(spelling), SearchContentType.Dictionary, new ModuleResultPolicy()),
            new(new SettingsModule(), SearchContentType.SystemTools, new ModuleResultPolicy()),
            new(new TextStatisticsModule(), SearchContentType.TextTools, new ModuleResultPolicy())
        };
        standards.AddRange((personalModules ?? []).Select(module =>
            new RegisteredSearchModule(module, SearchContentType.TextTools, new ModuleResultPolicy())));
        _standardModules = standards;
    }

    public Task PrepareAsync()
    {
        return Task.Run(() => _applications.Prepare());
    }

    public async Task<IReadOnlyList<LauncherResult>> SearchAsync(BackgroundSearchRequest request)
    {
        var cancellationToken = request.CancellationToken;
        if (request.FileNavigationActive)
        {
            if (!request.EnabledContentTypes.Contains(SearchContentType.Files))
            {
                return [];
            }

            var descriptor = new ModuleDescriptor(
                "file-navigation",
                "文件导航",
                new HashSet<ModuleCapability> { ModuleCapability.LocalFileRead });
            var results = await Task.Run(
                () => _fileNavigation.Results(
                    request.Query,
                    request.ShowsHiddenFiles,
                    request.FileNavigationSort,
                    request.FileNavigationSortAscending,
                    request.FileNavigationFoldersFirst),
                cancellationToken);
            cancellationToken.ThrowIfCancellationRequested();
            return Sanitize(
                results,
                descriptor,
                new ModuleResultPolicy(
                    allowedCapabilities: new HashSet<ModuleCapability> { ModuleCapability.LocalFileRead }));
        }

        var tasks = new List<Task<IReadOnlyList<LauncherResult>>>();
        if (request.EnabledContentTypes.Contains(SearchContentType.Applications))
        {
            tasks.Add(Task.Run(
                () => SearchApplications(
                    request.Query,
                    request.ApplicationAliases,
                    request.CustomApplicationPaths,
                    cancellationToken),
                cancellationToken));
        }

        foreach (var registration in _standardModules.Concat(request.RequestModules))
        {
            if (!request.EnabledContentTypes.Contains(registration.ContentType))
            {
                continue;
            }

            var policy = registration.Policy;
            var descriptor = registration.Module.Descriptor;
            if (!policy.Permits(descriptor))
            {
                continue;
            }

            tasks.Add(Task.Run(
                () => SearchModule(registration.Module, request.Query, policy, cancellationToken),
                cancellationToken));
        }

        var completed = await Task.WhenAll(tasks);
        cancellationToken.ThrowIfCancellationRequested();
        return completed.SelectMany(results => results).ToList();
    }

    private IReadOnlyList<LauncherResult> SearchApplications(
        string query,
        IReadOnlyDictionary<string, string> aliases,
        IReadOnlyList<string> customApplicationPaths,
        CancellationToken cancellationToken)
    {
        try
        {
            cancellationToken.ThrowIfCancellationRequested();
            var descriptor = new ModuleDescriptor(
                "applications",
                "应用程序",
                new HashSet<ModuleCapability>
                {
                    ModuleCapability.LocalFileRead,
                    ModuleCapability.ApplicationLaunch
                });
            var results = _applications.Search(query, aliases, customApplicationPaths);
            cancellationToken.ThrowIfCancellationRequested();
            return Sanitize(
                results,
                descriptor,
                new ModuleResultPolicy(
                    allowedCapabilities: new HashSet<ModuleCapability>
                    {
                        ModuleCapability.LocalFileRead,
                        ModuleCapability.ApplicationLaunch
                    }));
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw;
        }
        catch
        {
            // One provider must never blank the whole result set.
            return [];
        }
    }

    private static async Task<IReadOnlyList<LauncherResult>> SearchModule(
        IYToolsModule module,
        string query,
        ModuleResultPolicy policy,
        CancellationToken cancellationToken)
    {
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeout.CancelAfter(TimeSpan.FromSeconds(2));
        try
        {
            var request = new ModuleSearchRequest(query, 40, timeout.Token);
            var results = await module.SearchAsync(request).WaitAsync(timeout.Token);
            cancellationToken.ThrowIfCancellationRequested();
            return results
                .Take(request.MaximumResults)
                .Select(result => policy.Sanitize(result, module.Descriptor))
                .Where(result => result is not null)
                .Cast<LauncherResult>()
                .ToList();
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw;
        }
        catch
        {
            return [];
        }
    }

    private static IReadOnlyList<LauncherResult> Sanitize(
        IReadOnlyList<LauncherResult> results,
        ModuleDescriptor descriptor,
        ModuleResultPolicy policy)
    {
        return results
            .Take(40)
            .Select(result => policy.Sanitize(result, descriptor))
            .Where(result => result is not null)
            .Cast<LauncherResult>()
            .ToList();
    }

}
